cmake/CheckTypeRepresentation.cmake

```diff
@@ -0,0 +1,251 @@
+# Distributed under the OSI-approved BSD 3-Clause License.  See accompanying
+# file Copyright.txt or https://cmake.org/licensing for details.
+
+#[=======================================================================[.rst:
+CheckTypeRepresentation
+-----------------------
+
+Check the bit representation of a type.
+
+.. command:: check_float_type_representation
+
+  .. code-block:: cmake
+
+    check_float_type_representation(<type> <var> [LANGUAGE <language>])
+
+  Check if the floating point type exists and determine its bit format.
+  ``<type>`` can be any floating point type supported by the compiler, but is
+  typically one of ``float``, ``double``, or ``long double``. If the type
+  exists, then the variable ``HAVE_<var>`` is set to true, otherwise it is set
+  to false. If the type exists, and ``check_float_type_representation()`` was
+  able to detect the bit representation of ``<type>``, then the variable
+  ``<var>`` is set to its name. If the type exists, but
+  ``check_float_type_representation()`` was not able to detect its type, then
+  the variable ``<var>`` is set to ``UNKNOWN``. If the type does not exist,
+  then ``<var>`` is set to an empty string.
+
+  If ``LANGUAGE`` is set, the specified compiler will be used to perform the
+  check. Acceptable values are ``C`` and ``CXX``.
+
+  ``check_float_type_representation()`` understands the following floating
+  point formats:
+
+  ``IEEE754_B16_LE``
+  ``IEEE754_B16_BE``
+
+    IEEE 754 binary half-precision floating point number. Includes a ``LE`` or
+    ``BE`` suffix, depending on whether the number is little endian or big
+    endian, respectively.
```

Comment:
The 16-bit checks should also include the `bfloat16` format

Reply:
Do we have any platforms that use `bfloat16`?

Reply:
Not ATM but the way this is set up I think is sufficient to easily extend to add `FLOAT_HP_BFLOAT_{BE,LE}` at a later date.